File: nreduce/benchmark/mergesort.java
import java.util.*;

public class mergesort
{
  static List<String> genlist(int max)
  {
    List<String> res = new ArrayList<String>();
    for (int i = 0; i < max; i++) {
      char[] chars = ((7*i)+"").toCharArray();
      char[] reversed = new char[chars.length];
      for (int c = 0; c < chars.length; c++)
        reversed[chars.length-1-c] = chars[c];
      res.add(new String(reversed));
    }
    return res;
  }

  static void printList(List<String> list)
  {
    for (String s : list)
      System.out.println(s);
  }

  static List<String> mergesort(List<String> m)
  {
    if (m.size() <= 1)
      return m;

    int middle = m.size()/2;
    List<String> left = m.subList(0,middle);
    List<String> right = m.subList(middle,m.size());

    left = mergesort(left);
    right = mergesort(right);

    return merge(left,right);
  }

  static List<String> merge(List<String> xlst, List<String> ylst)
  {
    int x = 0;
    int y = 0;
    List<String> res = new ArrayList<String>();
    while ((x < xlst.size()) && (y < ylst.size())) {
      String xval = xlst.get(x);
      String yval = ylst.get(y);
      if (xval.compareTo(yval) < 0) {
        res.add(xval);
        x++;
      }
      else {
        res.add(yval);
        y++;
      }
    }
    if (x < xlst.size())
      res.addAll(xlst.subList(x,xlst.size()));
    if (y < ylst.size())
      res.addAll(ylst.subList(y,ylst.size()));
    return res;
  }

  public static void main(String[] args)
  {
    int n = 1000;
    if (args.length > 0)
      n = Integer.parseInt(args[0]);
    List<String> input = genlist(n);
    List<String> sorted = mergesort(input);
    printList(sorted);
  }

}

File: evaluation/benchmark/mergesort.py
#!/usr/bin/python

import sys

def genlist(max):
  res = []
  for i in range(0,max):
    rev = str(7*i)[::-1]
    res.append(rev)
  return res

def mergesort(m):
  if len(m) <= 1:
    return m

  middle = len(m)/2
  left = m[0:middle]
  right = m[middle:]

  left = mergesort(left)
  right = mergesort(right)

  return merge(left,right)

def merge(xlst,ylst):
  x = 0
  y = 0
  res = []
  while (x < len(xlst)) and (y < len(ylst)):
    xval = xlst[x]
    yval = ylst[y]
    if xval < yval:
      res.append(xval)
      x = x+1
    else:
      res.append(yval)
      y = y+1

  if x < len(xlst):
    res += xlst[x:]
  if y < len(ylst):
    res += ylst[y:]

  return res

n = 1000
if (len(sys.argv) >= 2):
  n = int(sys.argv[1])
input = genlist(n)
sorted = mergesort(input)
for s in sorted:
  print s
